spec/index.bs

```diff
@@ -2517,6 +2561,112 @@ dictionary GPUImageBitmapCopyView {
+
+<div algorithm="GPUCommandEncoder.copyBufferToTexture">
+
+  **Arguments:**
+    - {{GPUBufferCopyView}} |source|
+    - {{GPUTextureCopyView}} |destination|
+    - {{GPUExtent3D}} |copySize|
+
+  **Returns:** void
+
+  Encode a command into the {{GPUCommandEncoder}} that copies data from a sub-region of a
+  {{GPUBuffer}} to a sub-region of a {{GPUTexture}} [=subresource=].
+
+  |source| and |copySize| define the region of the source buffer.
+
+  |destination| and |copySize| define the region of the destination texture [=subresource=].
+
+</div>
+
+<div algorithm class=validusage>
+
+<dfn abstract-op>copyBufferToTexture Valid Usage</dfn>
+
+Given a {{GPUCommandEncoder}} |encoder| and the arguments {{GPUBufferCopyView}} |source|,
+{{GPUTextureCopyView}} |destination|, {{GPUExtent3D}} |copySize|, let
+  - |blockWidth| be the [=texel block width=] of |destination|.{{GPUTextureCopyView/texture}}.{{GPUTexture/[[format]]}}.
+  - |blockHeight| be the [=texel block height=] of |destination|.{{GPUTextureCopyView/texture}}.{{GPUTexture/[[format]]}}.
+  - |blockSize| be the [=texel block size=] of |destination|.{{GPUTextureCopyView/texture}}.{{GPUTexture/[[format]]}}.
+  - |bytes in a complete row| be |blockSize| &times; |copySize|.[=Extent3D/width=] &div; |blockWidth|.
+  - |required bytes in copy| be calculated with the following algorithm assuming all the parameters are [=valid=]. An error
+    occurs when there is any overflow in the calculation of |required bytes in copy|:
+    ```
+    GPUSize64 bytesPerRowOfTexelBlocks = (source.bytesPerRow == 0) ? bytesInACompleteRow : source.bytesPerRow;
+    GPUSize64 texelBlockRowsPerImage =
+        (source.rowsPerImage == 0) ? (copySize.height / blockHeight) : (source.rowsPerImage / blockHeight);
+    GPUSize64 bytesPerImage = bytesPerRowOfTexelBlocks * texelBlockRowsPerImage;
+    GPUSize64 bytesInLastSlice =
+        bytesPerRowOfTexelBlocks * (copySize.height / blockHeight - 1) + (copySize.width / blockWidth * blockSize);
+    requiredBytesInCopy = bytesPerImage * (copySize.depth - 1) + bytesInLastSlice;
+    ```
+
+  The following validation rules apply:
+
+  For |encoder|:
+  - |encoder|.{{GPUCommandEncoder/copyBufferToTexture()}} must not be called when a {{GPURenderPassEncoder}}
+    is active on |encoder|.
+  - |encoder|.{{GPUCommandEncoder/copyBufferToTexture()}} must not be called when a {{GPUComputePassEncoder}}
+    is active on |encoder|.
+
+  For |source|:
+  - |source| must be [=valid=].
+  - |source|.{{GPUBufferCopyView/buffer}}.{{GPUBuffer/[[usage]]}} must contain {{GPUBufferUsage/COPY_SRC}}.
+
+  For |destination|:
+  - |destination| must be [=valid=].
+  - |destination|.{{GPUTextureCopyView/texture}}.{{GPUTexture/[[textureUsage]]}} must contain
+    {{GPUTextureUsage/COPY_DST}}.
+
+  For |copySize|:
+  - |copySize|.[=Extent3D/width=] cannot be 0.
+  - |copySize|.[=Extent3D/height=] cannot be 0.
+  - |copySize|.[=Extent3D/depth=] cannot be 0.
```

Comment:
I found on D3D12 it is not allowed to specify copySize.width, copySize.height or copySize.depth to 0 when I was writing SPEC for copy commands, so here I suggest we not allow copySize.width, copySize.height or copySize.depth be 0.

Here are the error message reported from D3D12 validation layer:

D3D12 ERROR: ID3D12CommandList::CopyTextureRegion: D3D12_SUBRESOURCE_FOOTPRINT::Width, D3D12_SUBRESOURCE_FOOTPRINT::Height, and D3D12_SUBRESOURCE_FOOTPRINT::Depth cannot be 0. The minimum value is 1. Width = 64, Height = 0, and Depth = 1. [ RESOURCE_MANIPULATION ERROR #854: COPYTEXTUREREGION_INVALIDDSTDIMENSIONS]

Reply:
Imho empty copies should be allowed, because making them noops is as much work as validating them out.

Reply:
So you are suggesting that "it is no-op if any of the copySize.width, copySize.height or copySize.depth is 0"?

Reply:
Yes, but that can be non-normative ("Note: ...") because when we actually define the copies they'll probably be done as for-loops which describe that behavior.